### ISSUE TITLE
JETSCAPE doesn't catch empty initiator list

### ISSUE THREAD

**Issue by kkauder** [JETSCAPE/JETSCAPE#33]
For context, I generate some unusual "hard" processes where no hard partons are produced, and I still want to pass unstable hadrons etc. through the machinery. However, if a HardProcess simply doesn't call AddParton for some event, JETSCAPE silently uses one from the previous event. If it's never been called, a Segmentation Violation ensues, not elegant but at least obvious. But if previous events worked, there's no indication that the current event uses a made up one.
To reproduce: E.g. wrap lines 292+ from PythiaGun.cc in     
```
if ( GetCurrentEvent()<5 ){
}
``` 
and produce ten events. Result:
```
4 Event
...
# HardProcess Parton List: PythiaGun
0 21 0 55.3398 0.32149 0.343889 58.2244 0 3.3 0 0
0 21 0 53.1524 -1.92974 4.10431 186.907 0 3.3 0 0
...
# Energy loss Shower Initating Parton: JetEnergyLoss
0 21 0 55.3398 0.32149 0.343889 58.2244 0 3.3 0 0

5 Event
...
# HardProcess Parton List: PythiaGun
# Energy loss Shower Initating Parton: JetEnergyLoss
0 21 0 55.3398 0.32149 0.343889 58.2244 0 3.3 0 0
```

and it stays like this for the remaining events.

I think I can circumvent this problem by for example forcing a very soft initiator in the HardProcess, but this situation should be caught at the framework level and points to insufficient clearage of data from previous events.

**Comment by jhputschke** [comment]
Sorry, overlooked that issue you posted. Interesting, guess thats a case in which smart pointers causes the issues here. You are correct, one way would be indeed to explicitly clean up and throw an exception for example ...

**Comment by jdmulligan** [comment]
Hi @kkauder,

Thanks for pointing this out :) And sorry for the delay.
The problem is actually due to an (unprotected) out-of-bounds std::vector access in JetEnergyLossManager:

```
dynamic_pointer_cast<JetEnergyLoss>(it)->AddShowerInitiatingParton(hp[n])
```

which gives unpredictable behavior if `hp` is accessed out of bounds. 

The proper fix is simply to use .at() instead of []. (which should always be done ;) )
We'll fix it for the upcoming release.

It's also straightforward to allow 0 shower initiating partons without crashing -- we can add that also in the release. For example -- in case it's useful to you in the meantime:

- In `JetEnergyLoss::Clear()`: add `inP = nullptr;`
- In `JetEnergyLossManage::Exec()`, for the loop with `AddShowerInitiatingParton()`: change the if statement to `if (GetGetHardPartonListConnected() && hp.size() > 0)`

Cheers,
James

**Comment by jdmulligan** [comment]
Fixed in new release JETSCAPE 3.0
